### PR TITLE
Describing how to determine position from file line

### PR DESCRIPTION
**Pull request by tomasbjerre**

### CODE REVIEW

**Comment by bswinnerton** [comment]
👋 Hi @tomasbjerre, sorry for the delay on this. I'm conversing with some of my colleagues here at GitHub on the best sort of messaging to use in this doc. One of us will report back as soon as we can.

**Comment by bswinnerton** [comment]
For the sake of posterity, this pull request is a continuation of https://github.com/github/developer.github.com/pull/1020#discussion_r56004175.

**Comment by davidcelis** [comment]
Hey @tomasbjerre 👋 Sorry again for the delay. Thanks for getting the ball rolling with this clarifying note! We really like this idea of letting API users know how they can more easily calculate the `position` of the line for their comment, so how would you feel about expanding this comment with some more info? We were thinking something like…

> **Note**: To comment on a specific line, you will need to first determine the `position` from the diff. GitHub offers a [`application/vnd.github.v3.diff`](https://developer.github.com/v3/media/#diff) media type which you can use in a preceding request to view the pull request's diff. The `position` value is the number of lines down from the first "`@@`" hunk header in the file you would like to comment on. The line just below the "`@@`" line is `position` 1, the next line is `position` 2, and so on.

This is also the kind of thing I think we might want in a noticeable blue bubble, so we'd probably want to wrap it in a `{{#tip}} … {{/tip}}` block 😄 One last thing though: those tip blocks have picky whitespace requirements; there needs to be a newline separating the comment from the opening and closing tags, so it would look like:

```
{{#tip}}

<comment>

{{/tip}}
```

What do you think? Thanks again for taking the time to contribute to our docmentation 🙇

**Comment by Shredder121** [comment]
> One last thing though: those tip blocks have picky whitespace requirements; there needs to be a newline separating the comment from the opening and closing tags

Haha, yup. I can second that :+1: :wink:

Is it maybe a good idea to also mention the `diff_url` property of the pull request?
I know it's probably the same idea, but not everyone wants the same workflow, and maybe this gives end users more ideas on how to use the API?

**Comment by davidcelis** [comment]
Thanks for the suggestion, @Shredder121! It's probably not obvious, but because the `diff_url` goes through a redirect to the githubusercontent CDN, we believe that staying within the API and using the `application/vnd.github.v3.diff` media type is the best way for API clients to calculate the `position` for leaving comments.

**Comment by Shredder121** [comment]
Is it possible to add a section on how to count `position`s on multiple diff hunks?
I know how to read hunk headers, but have no idea what the position is of the second hunk.
[This PR ](https://github.com/github/developer.github.com/pull/1029)for example, [with diff](https://patch-diff.githubusercontent.com/raw/github/developer.github.com/pull/1029.diff).
Does the second hunk continue counting where the other left off?

I suppose that writing something [like this StackOverflow answer on the hunk header](http://stackoverflow.com/a/24456418/3485069) might be exactly what we need.

**Comment by davidcelis** [comment]
> Does the second hunk continue counting where the other left off?

Yup. The `position` simply keeps incrementing for each line in the diff regardless of whether a new hunk has begun or if the line is whitespace.

> [This PR](https://github.com/github/developer.github.com/pull/1029) for example, [with diff](https://patch-diff.githubusercontent.com/raw/github/developer.github.com/pull/1029.diff).

As an example here… If I wanted to leave a comment on that PR right at the line that reads `-[github-api-v3-wrapper]: https://github.com/Scion-Framework/GitHubAPI`, I would submit with a `position` of 13.

> Is it possible to add a section on how to count positions on multiple diff hunks?

Because of the above, I don't think this should be necessary. It should be enough to say "starting from the first `@@` hunk header."

**Comment by Shredder121** [comment]
>As an example here… If I wanted to leave a comment on that PR right at the line that reads -[github-api-v3-wrapper]: https://github.com/Scion-Framework/GitHubAPI, I would submit with a position of 13.

>>Is it possible to add a section on how to count positions on multiple diff hunks?

>Because of the above, I don't think this should be necessary. It should be enough to say "starting from the first @@ hunk header."

Okay, I'm just saying that we could inflate the text a bit.
Not too much, it shouldn't become the next guide, but IMO it having a bit more content is nice.

As an aside, @tomasbjerre I like the link you added!

Anyway, I leave the decision up to you @davidcelis, or others from GitHub for that matter, but maybe we could add a concrete example:

You set up a linter that has GitHub integration and adds comments when there is something wrong.
I'll use [this file for convenience][unified diff], since it's the same that's linked currently.

The linter says `'new'[28] - file doesn't end in a newline`.

So we would know that line 28 is contained in `[22, 22+7)`, so the lengths of the preceding hunks should be summed (`9+10 = 19`) and the final offset added (`19+7 = 26`).

If I recount the positions, then I indeed end up at 26, can you confirm that?
Does it make sense to have something in writing like this?


[unified diff]: https://en.wikipedia.org/wiki/Diff_utility#Unified_format

**Comment by davidcelis** [comment]
> Does the second hunk continue counting where the other left off?

Yup. The `position` just keeps incrementing for each line in the diff regardless of whether a new hunk has begun or if the line is whitespace (or even an additional hunk header).

> [This PR](https://github.com/github/developer.github.com/pull/1029) for example, [with diff](https://patch-diff.githubusercontent.com/raw/github/developer.github.com/pull/1029.diff).

As an example here… If I wanted to leave a comment on that PR right at the line that reads `-[github-api-v3-wrapper]: https://github.com/Scion-Framework/GitHubAPI`, I would submit with a `position` of 13.

> Is it possible to add a section on how to count positions on multiple diff hunks?

Because of the above, I don't think this should be necessary. It should be enough to say "starting from the first `@@` hunk header."

**Comment by davidcelis** [comment]
My comment vanished, I think I accidentally deleted it; so I reposted it above 😟 

What if we added a line to the very end: "The `position` in the file's diff continues to increase through lines of whitespace and additional hunks until a new file is reached."

**Comment by Shredder121** [comment]
Sounds good to me, @tomasbjerre what do you think?

**Comment by tomasbjerre** [comment]
Sure, I updated the PR.

**Comment by davidcelis** [comment]
Thanks @tomasbjerre, this is looking really good. Additionally, thanks for adding that `tip` block to the above note!

The one thing I'm wondering now is about having two tip blocks right next to each other. I think that we could probably just combine them into one and get rid of the "Note:" prefixes:

```
{{#tip}}

When passing the `commit_id`, use the SHA of the latest commit in the pull request or your comment may appear as "outdated" if the specified `position` has been modified in a subsequent commit.

To comment on a specific line, in a file, you will need to first determine the position, in the diff. GitHub offers a `application/vnd.github.v3.diff` media type which you can use in a preceding request to view the pull request's diff. The diff needs to be [interpreted](https://en.wikipedia.org/wiki/Diff_utility#Unified_format) to translate *line in file* to *position in diff*. The `position` value is the number of lines down from the first "@@" hunk header in the file you would like to comment on. The line just below the "@@" line is position 1, the next line is position 2, and so on. The position in the file's diff continues to increase through lines of whitespace and additional hunks until a new file is reached.

{{/tip}}
```

We should be good to go once you make that change. Thanks for all your patience 😄

**Comment by Shredder121** [comment]
@davidcelis This is slightly off-topic, but not too much.
Is it possible to add an inline comment on the left side (the old file) via the API?
That doesn't seem to be documented anywhere.

**Comment by Shredder121** [comment]
I mean, in unified diff mode it's not quite left-right, but when viewing it in split diff on the Web UI, the right or left side could be some additional context on what is being commented on.

File: content/v3/pulls/comments.md [hunk 1]
```diff
@@ -66,8 +66,13 @@ Name | Type | Description
...
 
-Note: When passing the `commit_id`, use the SHA of the latest commit in the pull request or your comment may appear as "outdated" if the specified `position` has been modified in a subsequent commit.
+{{#tip}}
 
+When passing the `commit_id`, use the SHA of the latest commit in the pull request or your comment may appear as "outdated" if the specified `position` has been modified in a subsequent commit.
+
+To comment on a specific line, in a file, you will need to first determine the position, in the diff. GitHub offers a `application/vnd.github.v3.diff` media type which you can use in a preceding request to view the pull request's diff. The diff needs to be [interpreted](https://en.wikipedia.org/wiki/Diff_utility#Unified_format) to translate *line in file* to *position in diff*. The `position` value is the number of lines down from the first "@@" hunk header in the file you would like to comment on. The line just below the "@@" line is position 1, the next line is position 2, and so on. The position in the file's diff continues to increase through lines of whitespace and additional hunks until a new file is reached.
```
Comment (davidcelis) [inline content/v3/pulls/comments.md]: Sorry, I found just a couple of nitpicks… There are a couple commas here that aren't necessary; could we get rid of the following commas?

* "To comment on a specific line, in a file"
* "you will need to first determine the `position`, in the diff"

Also, I noticed a couple things to help match this more with the style of the rest of our documentation:

* Changing "to translate *line in file*" to "to translate from the *line in the file*"
* Changing "to *position in diff*" to "to a *position in the diff*"

That should be it 😅

**Comment by davidcelis** [comment]
> @davidcelis This is slightly off-topic, but not too much. Is it possible to add an inline comment on the left side (the old file) via the API? That doesn't seem to be documented anywhere. I mean, in unified diff mode it's not quite left-right, but when viewing it in split diff on the Web UI, the right or left side could be some additional context on what is being commented on.

Not too off topic 😄  The split diff isn't used when submitting comments via the API. If the line you're commenting on is a deletion, that comment will appear on the left. Likewise, if the line being commented on is an addition, it'll appear on the right.

**Comment by Shredder121** [comment]
> If the line you're commenting on is a deletion, that comment will appear on the left. Likewise, if the line being commented on is an addition, it'll appear on the right

Oh. So the `position` count counts *all* the lines? That's really unintuitive.
But okay, I can see it now.
I'll try it out someday myself via the API anyway.
Thanks for answering.

**Comment by davidcelis** [comment]
> Oh. So the position count counts all the lines? That's really unintuitive.

Yes, sorry if that's confusing. We're trying to make it more clear in the documentation that `position` refers to the position in the _diff_ and not the file.

**Comment by davidcelis** [comment]
Thank you so much for taking the time to contribute, @tomasbjerre! ✨

**Pull request merged**